README.org
* Soundcloud Syncer
[[http://travis-ci.org/Sliim/soundcloud-syncer][http://secure.travis-ci.org/Sliim/soundcloud-syncer.png?branch=master]]

Soundcloud-syncer is a tool to synchronize soundcloud user's favorites and tracks locally.

** Requirements
*** Python version
1. Python 3.2.
*** Run unittests
To run unitests you will need following packages:
- mock

Optionnaly you can install these packages:
- nose
- coverage

#+BEGIN_SRC shell
pip-3.2 install mock nose coverage
#+END_SRC
** Usage
- Clone repository:
#+BEGIN_SRC shell
git clone https://github.com/Sliim/soundcloud-syncer.git
#+END_SRC

- Move into cloned directory and run ~sync.py~ script:
#+BEGIN_SRC shell
mv soundcloud-syncer
./sync.py --help
#+END_SRC

*** Basic usage
- Download user's likes:
#+BEGIN_SRC shell
./sync.py -u USER_TO_SYNC -o OUTPUT_DIRECTORY
#+END_SRC

- Download user's tracks:
#+BEGIN_SRC shell
./sync.py -u USER_TO_SYNC -o OUTPUT_DIRECTORY -t
#+END_SRC

- You can download tracks recursively:
#+BEGIN_SRC shell
./sync.py -u USER_TO_SYNC -o OUTPUT_DIRECTORY -t -r
#+END_SRC

- To set offset and limit for tracks list:
#+BEGIN_SRC shell
./sync.py -u USER_TO_SYNC -o OUTPUT_DIRECTORY -L LIMIT -O OFFSET
#+END_SRC
/!\ Limit can't be greater than 200.

** Ignoring tracks
You can ignore some tracks you don't want to sync locally.

To do this create a file in OUTPUT_DIRECTORY named ~.ignore~.
Each track in this file will be ignored.

- Example:
To ignore https://soundcloud.com/trapmusic/vexed-goodbye-by-j-nitrous add in ~.ignore~ file:
#+BEGIN_SRC txt
trapmusic/vexed-goodbye-by-j-nitrous
#+END_SRC
** Features
- Download user's favorites / likes tracks list.
- Download user's tracks list.
- Possibility to set offset and limit for tracks list.
- Auto find client_id when not passed to sync script.
- Check if track exists locally before download it.
- Possibility to ignore some tracks to sync.
- Recursive download.
** License
   See COPYING file


CHANGELOG
2013-03-07 Sliim <sliim@mailoo.org>
	* *.py:
	First tool release.


run-tests.sh
#!/bin/bash
#
# To run tests you need install nose, coverage and mock packages:
# pip-3.2 install nose coverage mock
#
nosetests-3.2 \
    -w tests/ssyncer \
    --with-coverage \
    --cover-package=ssyncer \
    --cover-erase \
    --cover-html


tests/ssyncer/test_suser.py
# This source file is part of Soundcloud-syncer.
#
# Soundcloud-syncer is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Soundcloud-syncer is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# Soundcloud-syncer. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.

import sys
import unittest

from mock import Mock
from mock import MagicMock

sys.path.insert(0, "../../")
from ssyncer.suser import suser
from ssyncer.strack import strack


def mock_tracks_response(uri):
    def json_res():
        return b'[{"kind":"track","id":1337,"title":"Foo","permalink":"foo","downloadable":true,"user":{"permalink":"user1"}, "original_format":"mp3"},{"kind":"track","id":1338,"title":"Bar","permalink":"bar","downloadable":true,"user":{"permalink":"user2"}, "original_format":"mp3"},{"kind":"track","id":1339,"title":"Baz","permalink":"baz","downloadable":true,"user":{"permalink":"user3"}, "original_format":"wav"}]'

    response = MagicMock()
    response.read = json_res
    return response


class TestSuser(unittest.TestCase):

    def test_object_has_good_name(self):
        """ Test object has name `foo`. """
        client = Mock()
        object = suser("Foo", client=client)
        self.assertEqual("Foo", object.name)

    def test_parse_tracks_response(self):
        """ Test parse tracks response on success. """
        client = MagicMock()
        object = suser("Foo", client=client)
        tracks = object._parse_tracks_response(mock_tracks_response("bar"))
        self.assertEquals(3, len(tracks))
        for track in tracks:
            self.assertIsInstance(track, strack)

    def test_get_likes_with_default_offset_and_limit(self):
        """ Test get user's likes with default offset and limit. """
        client = MagicMock()
        client.USER_LIKES = "/u/%s/f.json?o=%d&l=%d&c="
        object = suser("Foo", client=client)
        object._parse_tracks_response = Mock()
        object.get_likes()
        client.get.assert_called_once_with(
            "/u/Foo/f.json?o=0&l=50&c=")

    def test_get_likes_with_custom_offset_and_limit(self):
        """ Test get user's likes with custom offset and limit. """
        client = MagicMock()
        client.USER_LIKES = "/u/%s/f.json?o=%d&l=%d&c="
        object = suser("Foo", client=client)
        object._parse_tracks_response = Mock()
        object.get_likes(10, 20)
        client.get.assert_called_once_with(
            "/u/Foo/f.json?o=10&l=20&c=")

    def test_get_tracks_with_default_offset_and_limit(self):
        """ Test get user's tracks with default offset and limit. """
        client = MagicMock()
        client.USER_TRACKS = "/u/%s/t.json?o=%d&l=%d&c="
        object = suser("Foo", client=client)
        object._parse_tracks_response = Mock()
        object.get_tracks()
        client.get.assert_called_once_with(
            "/u/Foo/t.json?o=0&l=50&c=")

    def test_get_tracks_with_custom_offset_and_limit(self):
        """ Test get user's tracks with custom offset and limit. """
        client = MagicMock()
        client.USER_TRACKS = "/u/%s/t.json?o=%d&l=%d&c="
        object = suser("Foo", client=client)
        object._parse_tracks_response = Mock()
        object.get_tracks(10, 20)
        client.get.assert_called_once_with(
            "/u/Foo/t.json?o=10&l=20&c=")


sync.py
#!/usr/bin/env python3.2
# This source file is part of Soundcloud-syncer.
#
# Soundcloud-syncer is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Soundcloud-syncer is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# Soundcloud-syncer. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.

import argparse
import os.path

from ssyncer.sclient import sclient
from ssyncer.suser import suser
from ssyncer.serror import serror

parser = argparse.ArgumentParser(description="Soundcloud Syncer")
parser.add_argument("-u", "--user", help="Soundcloud user to sync", type=str, required=True)
parser.add_argument("-c", "--client-id", help="Your client id", type=str)
parser.add_argument("-o", "--output-dir", help="Output directory", type=str, required=True)
parser.add_argument("-O", "--offset", help="Tracks offset", type=int)
parser.add_argument("-L", "--limit", help="Tracks limit (max: 200)", type=int)
parser.add_argument("-r", "--recursive", help="Recursive download", action="store_true")
parser.add_argument("-t", "--tracks", help="Download user's tracks instead user's likes", action="store_true")


def downloader(tgetter, output, *args):
    """ Download tracks. Return number of tracks downloaded. """
    try:
        tracks = tgetter(*args)
    except serror as e:
        print(e)
        return False

    for strack in tracks:
        try:
            strack.download(output)
        except serror as e:
            print(e)

    return len(tracks)

args = parser.parse_args()

if not os.path.exists(args.output_dir):
    print(serror("Error: output directory `%s` doesn's exists." %
                 args.output_dir))
    exit(1)

offset = 0
limit = 50
if args.offset:
    offset = args.offset
if args.limit:
    if args.limit > 200:
        print(serror("Error: tracks limit limited to 200 tracks.."))
        exit(2)
    limit = args.limit

try:
    sclient = sclient(args.client_id)
    suser = suser(args.user, client=sclient)
except serror as e:
    print(e)
    exit(3)

tracks_getter = suser.get_likes
if args.tracks:
    tracks_getter = suser.get_tracks

if args.recursive:
    while True:
        nb = downloader(tracks_getter, args.output_dir, offset, limit)
        if nb == 0:
            break

        offset = offset + limit
else:
    downloader(tracks_getter, args.output_dir, offset, limit)


ssyncer/suser.py
# This source file is part of Soundcloud-syncer.
#
# Soundcloud-syncer is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Soundcloud-syncer is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# Soundcloud-syncer. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.

from ssyncer.sclient import sclient
from ssyncer.strack import strack
import json


class suser:

    name = None
    client = None

    def __init__(self, username, **kwargs):
        """ Initialize soundcloud's user object. """
        self.name = username

        if "client" in kwargs:
            self.client = kwargs.get("client")
        elif "client_id" in kwargs:
            self.client = sclient(kwargs.get("client_id"))
        else:
            self.client = sclient()

    def get_likes(self, offset=0, limit=50):
        """ Get user's likes. """
        response = self.client.get(
            self.client.USER_LIKES % (self.name, offset, limit))
        return self._parse_tracks_response(response)

    def get_tracks(self, offset=0, limit=50):
        """ Get user's tracks. """
        response = self.client.get(
            self.client.USER_TRACKS % (self.name, offset, limit))
        return self._parse_tracks_response(response)

    def _parse_tracks_response(self, response):
        """ Parse http response that contents tracks list. """
        objects = json.loads(response.read().decode("utf-8"))
        tracks = []

        for track in objects:
            tracks.append(strack(track, client=self.client))

        return tracks


ssyncer/serror.py
# This source file is part of Soundcloud-syncer.
#
# Soundcloud-syncer is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Soundcloud-syncer is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# Soundcloud-syncer. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.


class serror(Exception):
    """ Soundcloud-syncer errors exception. """

    message = ""

    def __init__(self, message):
        self.message = message

    def __str__(self):
        return "\033[91mERROR: %s\033[0m" % self.message


ssyncer/sclient.py
# This source file is part of Soundcloud-syncer.
#
# Soundcloud-syncer is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Soundcloud-syncer is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# Soundcloud-syncer. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.

import re
import urllib.request
from ssyncer.serror import serror


class sclient:
    host = "api.soundcloud.com"
    port = "443"
    client_id = None

    SC_HOME = "https://www.soundcloud.com"
    DOWNLOAD_URL = "/tracks/%d/download?client_id="
    STREAM_URL = "/tracks/%d/stream?client_id="
    USER_LIKES = "/users/%s/favorites.json?offset=%d&limit=%d&client_id="
    USER_TRACKS = "/users/%s/tracks.json?offset=%d&limit=%d&client_id="

    def __init__(self, client_id=None, **kwargs):
        """ Http client initialization. """
        if "host" in kwargs:
            self.host = kwargs.get("host")
        if "port" in kwargs:
            self.port = kwargs.get("port")

        if not client_id:
            print("INFO: Attempt to get client_id..")
            self.client_id = self.get_client_id()
            print("INFO: OK, client_id = %s" % self.client_id)
        else:
            self.client_id = client_id

    def get_protocol(self):
        """ Get protocol from port to use. """
        if self.port == "443":
            return "https"
        return "http"

    def send_request(self, url):
        """ Send a request to given url. """
        try:
            return urllib.request.urlopen(url)
        except urllib.error.HTTPError as e:
            raise serror(
                "Request `%s` failed (%s:%s)." %
                (url, e.__class__.__name__, e.code))

    def get(self, uri):
        """ Send a request to given uri. """
        return self.send_request(
            "{0}://{1}:{2}{3}{4}".format(
                self.get_protocol(),
                self.host,
                self.port,
                uri,
                self.client_id
            )
        )

    def get_location(self, uri):
        """ Send a request and get redirected url. """
        return self.get(uri).geturl()

    def get_client_id(self):
        """ Attempt to get client_id from soundcloud homepage. """
        id = re.search(
            "\"clientID\":\"([a-z0-9]*)\"",
            self.send_request(self.SC_HOME).read().decode("utf-8"))

        if not id:
            raise serror("Cannot retrieve client_id.")

        return id.group(1)


ssyncer/strack.py
# This source file is part of Soundcloud-syncer.
#
# Soundcloud-syncer is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Soundcloud-syncer is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# Soundcloud-syncer. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.

import sys
import os.path
import urllib.request

from ssyncer.sclient import sclient
from ssyncer.serror import serror


class strack:

    client = None
    metadata = {}

    def __init__(self, track_data, **kwargs):
        """ Track object initialization, load track metadata. """
        if "client" in kwargs:
            self.client = kwargs.get("client")
        elif "client_id" in kwargs:
            self.client = sclient(kwargs.get("client_id"))
        else:
            self.client = sclient()

        self.metadata = {
            "id": track_data["id"],
            "title": track_data["title"],
            "permalink": track_data["permalink"],
            "username": track_data["user"]["permalink"],
            "downloadable": track_data["downloadable"],
            "ext": track_data["original_format"],
        }

    def get(self, key):
        """ Get track metadata value from a given key. """
        if key in self.metadata:
            return self.metadata[key]
        return None

    def get_download_link(self):
        """ Get direct download link with soudcloud's redirect system. """
        url = None
        if not self.get("downloadable"):
            try:
                url = self.client.get_location(
                    self.client.STREAM_URL % self.get("id"))
            except serror as e:
                print(e)

        if not url:
            try:
                url = self.client.get_location(
                    self.client.DOWNLOAD_URL % self.get("id"))
            except serror as e:
                print(e)

        return url

    def gen_filename(self):
        """ Generate local filename for this track. """
        return "{0}-{1}.{2}".format(
            self.get("id"),
            self.get("permalink"),
            self.get("ext"))

    def gen_localdir(self, localdir):
        """
        Generate local directory where track will be saved.
        Create it if not exists.
        """
        directory = "{0}/{1}/".format(localdir, self.get("username"))
        if not os.path.exists(directory):
            os.makedirs(directory)
        return directory

    def track_exists(self, localdir):
        """ Check if track exists in local directory. """
        path = self.gen_localdir(localdir) + self.gen_filename()
        if os.path.exists(path) and os.path.getsize(path) > 0:
            return True
        return False

    def get_ignored_tracks(self, localdir):
        """ Get ignored tracks list. """
        ignore_file = "%s/.ignore" % localdir
        list = []
        if os.path.exists(ignore_file):
            f = open(ignore_file)
            ignored = f.readlines()
            f.close()

            for i in ignored:
                list.append("%s/%s" % (localdir, i.rstrip()))

        return list

    def download(self, localdir):
        """ Download a track in local directory. """
        local_file = self.gen_localdir(localdir) + self.gen_filename()

        if self.track_exists(localdir):
            print("INFO: Track {0} already downloaded, skipping!".format(
                self.get("id")))
            return False

        if local_file in self.get_ignored_tracks(localdir):
            print("\033[93mINFO: Track {0} ignored, skipping!!\033[0m".format(
                self.get("id")))
            return False

        dlurl = self.get_download_link()

        if not dlurl:
            raise serror("Can't download track_id:%d|%s" % (
                self.get("id"),
                self.get("title")))

        try:
            print("Start downloading %s (%d).." % (
                self.get("title"),
                self.get("id")))
            urllib.request.urlretrieve(dlurl, local_file, self._progress_hook)
        except:
            os.remove(local_file)
            raise

    def _progress_hook(self, blocknum, blocksize, totalsize):
        """ Progress hook for urlretrieve. """
        read = blocknum * blocksize
        if totalsize > 0:
            percent = read * 1e2 / totalsize
            s = "\r%5.1f%% %*d / %d" % (
                percent, len(str(totalsize)), read, totalsize)
            sys.stderr.write(s)

            if read >= totalsize:
                sys.stderr.write("\n")
        else:
            sys.stderr.write("read %d\n" % read)
